Feat/member property permissions

msadura opened the pull request:
https://app.charmverse.io/charmverse/page-717681698804959

Property permissions:
- returning only accessigble props from api
- api for creating / removing proeprty permission by role
- UI updates
- tests


https://user-images.githubusercontent.com/5198394/197764800-25e000a6-55c5-4e7f-8e31-6f147f0a3819.mp4

motechFR commented on prisma/migrations/20221024201445_member_property_permission/migration.sql:
```diff
@@ -0,0 +1,27 @@
+-- CreateEnum
```
General note re:migrations, if you iterated the data model multiple times in PR, its best to delete and recreate the migrations so we have a single migration file before merge

motechFR commented on prisma/schema.prisma:
```diff
@@ -811,3 +819,19 @@ model MemberPropertyValue {
+model MemberPropertyPermission {
+  id                            String                        @id @default(uuid()) @db.Uuid
+  userId                        String?                       @db.Uuid
+  spaceId                       String?                       @db.Uuid
```
Here, you're using the spaceId key to group permissions. This is a bit confusing for developers, as we might think it's an assignable group who can receive permissions.

It's better to drop this field, and lookup the space of the linked Member property

motechFR commented on lib/members/createMemberPropertyPermission.ts:
```diff
@@ -0,0 +1,20 @@
+import type { CreateMemberPropertyPermissionInput } from 'lib/members/interfaces';
+
+export function createMemberPropertyPermission (data: CreateMemberPropertyPermissionInput) {
+  return prisma.memberPropertyPermission.create({
```
Security best practice: Add a check and error so we can only assign permissions to users and roles who are in the same space as this property.

motechFR commented on lib/members/getAccessibleMemberPropertiesBySpace.tsx:
```diff
@@ -0,0 +1,103 @@
+          OR: [
+            // User has permission to view property
+            {
+              spaceId: { in: spaceIdQuery },
```
After removing spaceId from MemberPropertyPermissions, you can update this to MemberProperty {spaceId}

motechFR commented on lib/members/getAccessibleMemberPropertiesBySpace.tsx:
```diff
@@ -0,0 +1,103 @@
+}
+
+export function accessiblePropertiesByPermissionsQuery ({ spaceIds, userId }:
+  { spaceIds: string[], userId: string }): Prisma.MemberPropertyPermissionListRelationFilter {
```
Philosophically, there's only ever 1 possible spaceId for a specific set of MemberProperties, best to update this to single string

motechFR commented on components/members/components/MemberDirectoryProperties/MemberPropertiesSidebar.tsx:
```diff
@@ -160,25 +157,29 @@ export function MemberPropertySidebarItem ({
+                  }}
+                />
+              </Tooltip>
             )}
             <ConfirmDeleteModal
               title='Delete property'
```
I'm not sure we need to have a Modal when deleting the permission. The UI immediately changes to "Everyone at X space"

motechFR commented:
@msadura This looks great, all our core paradigms have been ported over to Member properties.

For security, can you please add integration tests for the create and delete endpoints?

We need a 201 (admin user) and 401 (non admin user) for each endpoint.

Finally, for the list endpoint, can you copy / paste the tests from the getAccessibleProperties and adapt them to the server setup? You covered all the key cases, so no edits required

motechFR commented on lib/members/getAccessibleMemberPropertiesBySpace.tsx:
```diff
@@ -0,0 +1,103 @@
+        // User must be a member of space
+        {
+          space: {
+            id: { in: spaceIdQuery },
```
Is there a case where spaceId is actually an array of more than 1? If so, we should rename it to spaceIds, otherwise drop the in: query and ensure consumers pass a single spaceId

motechFR reviewed (approved):
Approved, but plz remember integration tests ;)

msadura commented:
@motechFR PR is ready for re-review.

-removed unised stuff from permissions model
- added business logic check before creating permission
- business logic tests and integrations tests for listing props with permissions, creating / deleting permissions

motechFR commented on __integration-tests__/server/pages/api/spaces/[id]/members/properties/permissions.spec.ts:
```diff
@@ -0,0 +1,110 @@
+      .expect(401)).body as MemberPropertyPermissionWithRole;
+  });
+
+  it('should return error for user from other space when trying to delete space', async () => {
```
text correction for better clarity

should return error for user from other space when trying to delete **permissions inside** the space

motechFR commented on __integration-tests__/server/pages/api/spaces/[id]/members/properties/permissions.spec.ts:
```diff
@@ -0,0 +1,110 @@
+      .expect(401)).body as MemberProperty[];
+  });
+
+  it('should return error for user from other space', async () => {
```
should return error for user from other space when they try to create a permission

motechFR commented on __integration-tests__/server/pages/api/spaces/[id]/members/properties/permissions.spec.ts:
```diff
@@ -0,0 +1,110 @@
+  role1 = await generateRole({ spaceId: space.id, roleName: 'test role 1', createdBy: adminUser.id });
+});
+
+describe('GET /api/space/[id]/members/properties/permissions - Create and delete property permissions', () => {
```
I think you forgot to rename this when copying over from other file.

Also, the delete endpoints should be under a separate describe group, but same file is fine (following API structure)
